src/pycram/datastructures/property.py

```diff
@@ -0,0 +1,361 @@
+
+    def manage_io(self, func: Callable, *args, **kwargs) -> bool:
+        """
+        Manages the ReasoningResult of a property function and calls it. The success of the method will be passed to the
+        parent node while the reasoned parameters will be stored in the variables dictionary of the root node.
+
+        :param func: Property function to call
+        :param args: args to pass to the function
+        :param kwargs: keyword args to pass to the function
+        :return: result of the function
+        """
+        reasoning_result = func(*args, **kwargs)
+        self.root.variables.update(reasoning_result.reasoned_parameter)
+        self.result = reasoning_result
+        return reasoning_result.success
+
+    @property
+    def successful(self) -> bool:
+        """
+        Returns the result of the complete tree structure. This is done by iterating over the tree and checking if all
+        nodes return True.
+
+        :return: True if all nodes in the tree are True, False otherwise
+        """
+        success = self.result
+        for node in PreOrderIter(self.root):
+            success &= node.result.success
+        return success
+
+    @property
+    def resolved(self) -> bool:
+        """
+        Returns True if all nodes in the tree are resolved. This is done by iterating over the tree and checking if all
+        nodes are either a ResolvedProperty or a PropertyOperator.
+
+        :return: True if all nodes in the tree are resolved, False otherwise
+        """
+        res = True
+        for node in PreOrderIter(self.root):
+            res &= isinstance(node, (ResolvedProperty, PropertyOperator))
+        return res
+
+
+class PropertyOperator(Property):
+    """
+    Parent class for logical operators to combine multiple properties in a tree structure. This class adds methods to
+    use Properties as part of a tree structure. Furthermore, there is a method to simplify the tree structure by merging
+    Nodes of the same type.
+    """
+
+    def __init__(self, properties: List[Property]):
+        """
+        Initialize the PropertyOperator with a list of Properties. The parent of this node is None, therefore the node is
+        always the root of the tree.
+
+        :param properties: A list of properties to which are the children of this node
+        """
+        super().__init__(parent=None, children=properties)
+        self.result = True
+
+    def simplify(self):
+        """
+        Simplifies the tree structure by merging nodes of the same type. This is done by iterating over the tree and
+        merging nodes of the same type.
+
+        :return: Returns the root node of the tree
+        """
+        for node in PreOrderIter(self.root):
+            for child in node.children:
+                if type(node) is type(child):
+                    self.merge_nodes(node, child)
+        return self.root
+
+    @staticmethod
+    def merge_nodes(node1: Node, node2: Node) -> None:
```

Comment:
i find the word "merging" for stealing the children quite confusing

Reply:
But that's exacty what it does, it combines two nodes by merging them